### DISCUSSION TITLE
How to gracefully mix Markdown and latex tables?

### DISCUSSION THREAD

**Discussion by tbeason** [quarto-dev/quarto-cli#2052] [Q&A]
Markdown tables work well for simple tables. I also like the `@tbl-name` cross-reference ability provided by Quarto. However, I have some tables that are somewhat problematic to write using Markdown and have them look nice. I know I can just put the raw latex in the file, but then I lose the ability to reference it using the same syntax. Is there a way to hack things together so that I can still reference latex tables using the `@` syntax?

**Comment by rgaiacs** [comment]
@tbeason can you provide a example of somewhat problematic table that you have?

**Reply by tbeason** [reply]
The main thing that is missing for me in Markdown is variable row spacing. For example, if I am outputting regression estimates and standard errors as "grouped rows", typically I would end a standard error line in latex like this: `\\[2mm]` which adds a little space before the next row. See for example what `esttab` does in Stata: http://repec.org/bocode/e/estout/esttab.html

**Reply by tbeason** [reply]
Alternatively, if there was a way to *not* double space a particular table when it is converted from Markdown, that could also solve my particular problems right now. My document uses `linestretch: 1.5` and it seems this is also applied to all tables. If I could turn it off selectively for tables, that would work well enough for me (although I would prefer variable row spacing).

**Reply by rgaiacs** [reply]
@tbeason Consider the following example inspired by the esttab documentation:

```
df <- tibble(
  product = c("weight", "", "mpg", ""),
  price1 = c("1.747", "(5.49)", "-49.51", "(-0.57)"),
  price2 = c("3.465", "(5.49)", "21.85", "(0.29)")
)
```

When you render your Quarto document to PDF with LaTeX, `knitr` uses the default settings. This is equivalent to `df |> knitr::kable("latex")` and produces

```
\begin{tabular}{l|l|l}
\hline
product & price1 & price2\\
\hline
weight & 1.747 & 3.465\\
\hline
 & (5.49) & (5.49)\\
\hline
mpg & -49.51 & 21.85\\
\hline
 & (-0.57) & (0.29)\\
\hline
\end{tabular}
```

`knitr::kable` has a lot of options for LaTeX, see https://github.com/yihui/knitr/blob/f467b6b27e1fee71bf0baa52ba13e048b66a215d/R/table.R#L211. You can be define the options that you want in your Quarto document, for example `df |> knitr::kable("latex", linesep = c("", "[2mm]"))` will produce

```
\begin{tabular}{l|l|l}
\hline
product & price1 & price2\\
\hline
weight & 1.747 & 3.465\\
 & (5.49) & (5.49)\\
[2mm]
mpg & -49.51 & 21.85\\
 & (-0.57) & (0.29)\\
\hline
\end{tabular}
```

That is the output that you want.

**Reply by tbeason** [reply]
I do not think this really solves my problem. I know how to make and output the table in latex. I guess I offered up more gripe than I should have since you asked about the table. I don't think there is a good way to make the table using Markdown syntax, but that isn't such a concern because I am fluent (enough) in latex.

The real problem is that, in my document, I want to be able to refer to a latex table the same way that I can refer to a Markdown table: using the `@tbl-name` syntax. If I have to remember which tables I made in Markdown and which tables were in latex in order to reference them, it isn't useful. Below is a MWE. I have also tried using a computational block which spit out the latex table and setting `|# label:` and `|# tbl-cap:` options in this block, but that did not work (and doesn't highlight well here, so I've excluded it from MWE).

```
---
title: "Plots Demo"
author: "Norah Jones"
date: "5/22/2021"
format:
  pdf: default
keep-tex: true
---

## Table Label Tests

Can I refer to these tables? Let's see @tbl-first or @tbl-second.


| Default | Left | Right | Center |
|---------|:-----|------:|:------:|
| 12      | 12   |    12 |   12   |
| 123     | 123  |   123 |  123   |
| 1       | 1    |     1 |   1    |

: Test Table {#tbl-first}


{{< pagebreak >}}

\begin{table}[t]
\centering
\caption{\label{tbl-second}Again here.}
\begin{tabular}{rrrrl}
\toprule
Sepal.Length & Sepal.Width & Petal.Length & Petal.Width & Species\\
\midrule
5.1 & 3.5 & 1.4 & 0.2 & setosa\\
4.9 & 3.0 & 1.4 & 0.2 & setosa\\
4.7 & 3.2 & 1.3 & 0.2 & setosa\\
\addlinespace
4.6 & 3.1 & 1.5 & 0.2 & setosa\\
5.0 & 3.6 & 1.4 & 0.2 & setosa\\
5.4 & 3.9 & 1.7 & 0.4 & setosa\\
\addlinespace
4.6 & 3.4 & 1.4 & 0.3 & setosa\\
5.0 & 3.4 & 1.5 & 0.2 & setosa\\
4.4 & 2.9 & 1.4 & 0.2 & setosa\\
\addlinespace
4.9 & 3.1 & 1.5 & 0.1 & setosa\\
\bottomrule
\end{tabular}
\end{table}
```

**Comment by eeholmes** [comment]
I tried this out and had the same result. Nothing when I include a LaTeX table in a qmd (or Rmd) file. I even tried
```
:::{=latex}
latex table
:::
```

This makes no sense. Pandoc coverts simple LaTeX tables just fine to HTML. It is often used as a LaTeX to HTML coverter.
e.g. https://tex.stackexchange.com/questions/161677/latex-tables-in-pandoc-captions

@cscheid Any idea what's wrong? I don't think it's a Quarto specific issue as I saw it in R Markdown too. Pandoc knows how to convert LaTeX tables to HTML, but for some reason is not 'doing its magic' on them here. Usually it is pretty smart about dealing with LaTeX embedded in markdown.

**Reply by cscheid** [reply]
Can you share a full qmd file that shows what's going wrong? Thanks.

**Reply by eeholmes** [reply]
This rendered to html produces nothing except the title.

```
---
title: "Untitled"
format: 
  pdf:
    keep-tex: true
editor: visual
---

:::{=latex}
\begin{tabular}{ c c c }
 cell1 & cell2 & cell3 \\ 
 cell4 & cell5 & cell6 \\  
 cell7 & cell8 & cell9    
\end{tabular}
:::

```

But if we render to tex, we get a pdf that looks like this:
<img width="871" alt="image" src="https://user-images.githubusercontent.com/2545978/186329528-aba33154-4c6b-4f7c-9e93-8d2a64158da4.png">

We can run the test.tex file through Pandoc to get html
```
pandoc -s test.tex -o test.html
```
The Pandoc produced html file shows this:

<img width="884" alt="image" src="https://user-images.githubusercontent.com/2545978/186329214-c0fac8b0-17fe-4c16-b7da-13c63745aced.png">

So Pandoc does know how to turn a basic LaTeX table into html.

**Reply by eeholmes** [reply]
Another tex example

table.tex
```
\begin{tabular}{ c c c }
 cell1 & cell2 & cell3 \\ 
 cell4 & cell5 & cell6 \\  
 cell7 & cell8 & cell9    
\end{tabular}
```
Running
```
pandoc  table.tex -o table.html
```
produces
```
<table>
<tbody>
<tr class="odd">
<td style="text-align: center;">cell1</td>
<td style="text-align: center;">cell2</td>
<td style="text-align: center;">cell3</td>
</tr>
<tr class="even">
<td style="text-align: center;">cell4</td>
<td style="text-align: center;">cell5</td>
<td style="text-align: center;">cell6</td>
</tr>
<tr class="odd">
<td style="text-align: center;">cell7</td>
<td style="text-align: center;">cell8</td>
<td style="text-align: center;">cell9</td>
</tr>
</tbody>
</table>
```
So why doesn't Pandoc interpret LaTeX tables if it knows how? Strange.

**Reply by rgaiacs** [reply]
@eeholmes The way that Quarto works is

1. Load `.qmd`.
2. If `.qmd` need to be executed, Quarto will figure out the engine (knitr or Jupyter), convert the `.qmd` to `.Rmd` or `.ipynb`, execute the file and save the it as `.md` file.
3. Quarto will get the `.md` file and send to Pandoc.

Consider the minimal working example that you provided before:

```
---
title: "Untitled"
format: 
  pdf:
    keep-tex: true
editor: visual
---

:::{=latex}
\begin{tabular}{ c c c }
 cell1 & cell2 & cell3 \\ 
 cell4 & cell5 & cell6 \\  
 cell7 & cell8 & cell9    
\end{tabular}
:::
```

The output of `pandoc -f markdown -t latex mwe.qmd` is

``` 
\begin{tabular}{ c c c }
 cell1 & cell2 & cell3 \\ 
 cell4 & cell5 & cell6 \\  
 cell7 & cell8 & cell9    
\end{tabular}
```

The output of `pandoc -f markdown -t html mwe.md` is

``` 
<div class="{=latex}">

</div>
```

When you use `:::{=latex}`, you are telling Pandoc that the code is a LaTeX that Pandoc must preserve when converting the document to LaTeX. This feature should be limited to cases where Pandoc can't represent the LaTeX internally.

**Reply by cscheid** [reply]
Whew, we're deep into Pandoc weeds here, so let's first make sure we are all in the same page.

## Raw Block Syntax

> When you use :::{=latex}, you are telling Pandoc that the code is a LaTeX that Pandoc must preserve when converting the document to LaTeX. This feature should be limited to cases where Pandoc can't represent the LaTeX internally.

Here's the internal Pandoc representation of a couple of different documents:

````
$ cat raw-test.qmd
:::{=latex}
\begin{tabular}{ c c c }
 cell1 & cell2 & cell3 \\ 
 cell4 & cell5 & cell6 \\  
 cell7 & cell8 & cell9    
\end{tabular}
:::
$ pandoc -f markdown -t native raw-test.qmd
[Div ("",["{=latex}"],[])
 [RawBlock (Format "tex") "\\begin{tabular}{ c c c }\n cell1 & cell2 & cell3 \\\\ \n cell4 & cell5 & cell6 \\\\  \n cell7 & cell8 & cell9    \n\\end{tabular}"]]
````

Compare that to this:

````
$ cat raw-test-2.qmd
```{=latex}
\begin{tabular}{ c c c }
 cell1 & cell2 & cell3 \\ 
 cell4 & cell5 & cell6 \\  
 cell7 & cell8 & cell9    
\end{tabular}
```
$ pandoc -f markdown -t native raw-test-2.qmd
[RawBlock (Format "latex") "\\begin{tabular}{ c c c }\n cell1 & cell2 & cell3 \\\\ \n cell4 & cell5 & cell6 \\\\  \n cell7 & cell8 & cell9    \n\\end{tabular}"]
````

This internal representation takes a bit of work to read at first, but you can tell that if we use `:::{=latex}`, we're not getting exactly what we expected. The issue here is that the syntax for a "raw latex block" is

````
```{=latex}
...
```
````

instead of

````
:::{=latex}
...
:::
````

What @rgaiacs meant originally about this syntax being designed for "please send this block directly to the latex format and only the latex format" is true, but it holds for the code-block syntax, not the div syntax.

## Format Portability for Tables

Let's go back to the point of this discussion: how to get tables in multiple formats. What @eeholmes was referring to is (probably, pls correct me if I'm wrong!) this:

````
$ cat raw-test-3.qmd
\begin{tabular}{ c c c }
 cell1 & cell2 & cell3 \\ 
 cell4 & cell5 & cell6 \\  
 cell7 & cell8 & cell9    
\end{tabular}
$ pandoc -f markdown -t html raw-test-3.qmd

$ pandoc -f latex -t html raw-test-3.qmd
<table>
<tbody>
<tr class="odd">
<td style="text-align: center;">cell1</td>
<td style="text-align: center;">cell2</td>
<td style="text-align: center;">cell3</td>
</tr>
<tr class="even">
<td style="text-align: center;">cell4</td>
<td style="text-align: center;">cell5</td>
<td style="text-align: center;">cell6</td>
</tr>
<tr class="odd">
<td style="text-align: center;">cell7</td>
<td style="text-align: center;">cell8</td>
<td style="text-align: center;">cell9</td>
</tr>
</tbody>
</table>
$ pandoc -f latex -t markdown raw-test-3.qmd
  ------- ------- -------
   cell1   cell2   cell3
   cell4   cell5   cell6
   cell7   cell8   cell9
  ------- ------- -------
````

So pandoc can convert from LaTeX to markdown and html tables (!!).

This is what's truly powerful about Pandoc, if you've never seen it before. It not only generates many output formats, it _reads_ many input formats as well. However, it requires us telling pandoc that the input format is LaTeX, which is not the case in quarto. If you try the following, it will not work:

```
$ quarto render raw-test-3.qmd --to html
```

This will not generate the content, for the same reason that `pandoc -f markdown -t html raw-test-3.qmd` doesn't.

## "But none of this solves my problem!"

Sadly, no. Ultimately, I think that what @tbeason wants to do is currently not easy in Quarto. We have plans to improve our crossref infrastructure for 1.2. It's possible that we will end up in a place where we can add "raw references", but we won't commit to it because it's quite a complicated feature with a lot of moving parts.

**Reply by cscheid** [reply]
Followup, there's a related trick we _already_ do in quarto, which is that if you use HTML tables as input, we _will_ convert them to LaTeX for you. Currently, we only do this for HTML table inputs, but we might change that in the future.

**Reply by eeholmes** [reply]
@cscheid Thanks for the deep dive into Pandoc! Yes, you captured what I was getting at. The effect I showed was the same whether or not `:::{=latex}` is used. The point being, as you caught, that Pandoc knows how to convert LaTeX tables to html but for some reason it doesn't here. Same happens in Rmd. Not a bug, probably just not implemented?

My idea with using `:::{=latex}` was that I would "tell" Pandoc/Quarto/knitr? that this is LaTeX so maybe it would not get confused and would use the Pandoc latex to html convertor on that chunk when rendering to html. I had no idea if this would work but since `:::{=latex}` exists I thought maybe it was used in this way.

For example, I use `\begin{equation} ... \end{equation}` for my equations instead of `$$ ... $$`---because as an applied mathematician, I need the LaTeX equation environment or similar for the types of equations I write. I noticed that when I switched between the visual and source editor in RStudio, my equations were put in `:::{=latex} ... :::` chunks. I concluded that Pandoc uses that latex blocking to identify when it needs to use its latex to html converter on a div. But that doesn't seem to be what is happening. How does Pandoc know to convert `\begin{equation} ... \end{equation}` to html but does not know to convert `\begin{tabular} ... \end{tabular}`?

**Reply by cscheid** [reply]
> Not a bug, probably just not implemented?

That's right. The issue is that Pandoc needs to know that the _input_ is latex to convert correctly. And, in quarto, our input is always Markdown.

> My idea with using :::{=latex} was that I would "tell" Pandoc/Quarto/knitr?

Yes, but what you're telling Pandoc is that whatever is inside this block, it should appear only when the _output_ is LaTeX. I think you're thinking of it as "Pandoc: here's some LaTeX input, do the right thing with it now". And, to be clear, when you use the `{=latex}` syntax, Pandoc will not touch it at all, and simply will insert the content as is into LaTeX output documents (as well as ignoring it when the output is _not_ a LaTeX document).

> How does Pandoc know to convert \begin{equation} ... \end{equation} to html but does not know to convert \begin{tabular} ... \end{tabular}?

I'm not 100% sure. But I think it's simply because Pandoc treats `equation` environments differently, and chooses to interpret the output as a supported way to typeset math.

**Comment by eeholmes** [comment]
Hmm, it seems to me that `:::{=latex}` has no effect. If Pandoc is going to interpret latex, it will do so whether or not it is in a `:::{=latex}` block. And if it is going to ignore that latex, it will ignore it whether or not it is in a `:::{=latex}` block.

Both of these render fine to html and the `:::{=latex}` bit has no effect. html output is the same whether it is there or not.

```
:::{=latex}
\begin{equation}
\sum a
\end{equation}
:::

:::{=latex}
\newcommand\muy{\mu_y}
:::

We use $\muy$ to denote the mean of the $y$ parameter
```

But yes, I would like `:::{=latex}` to mean "Pandoc: here's some LaTeX input, do the right thing with it now" but to me it looks Pandoc reads it as "I am going to ignore that completely and do whatever I was going to do anyhow."

Note the reason I find this interesting is that if a block like `:::{=some-format-that-pandoc-understands}` worked as "Pandoc: here's some ABC input that you already know, do the right thing with it now". Then that opens up all kinds of new things I could do. Whatever Pandoc understands would be a valid input and I could seamlessly convert to different formats.

**Comment by cscheid** [comment]
> But yes, I would like :::{=latex} 

I want to come back to `:::{=latex}` instead of 

````
```{=latex}
````

How are you getting these `:::{=latex}` blocks in your documents?

> ... to mean "Pandoc: here's some LaTeX input, do the right thing with it now" but to me it looks Pandoc reads it as "I am going to ignore that completely and do whatever I was going to do anyhow."

Unfortunately, we don't control that syntax; Pandoc does.

> Note the reason I find this interesting is that if a block like :::{=some-format-that-pandoc-understands} worked as "Pandoc: here's some ABC input that you already know, do the right thing with it now". Then that opens up all kinds of new things I could do. Whatever Pandoc understands would be a valid input and I could seamlessly convert to different formats.

Can you tell me about some of these new things? We've been thinking about a similar feature but we could only think of tables as a valid use case for it.

**Reply by eeholmes** [reply]
Ok, maybe 
<pre>
```{=latex}
...
```
</pre>
makes more sense for this application. Even if currently it means "ignore unless latex", that doesn't have to be the interpretation.

Regarding other applications. Well Pandoc can translate to-from all kinds of things: https://pandoc.org/

1) I could use content written in all sorts of formats wo worrying about how to convert to markdown. Just let Pandoc do it for me. All I have to do is tell Pandoc what it is seeing.

2) I could use any program that outputs in a particular format. Run a shell script within a code chunk and tell Pandoc what output is going to be. I don't have to worry about there being a xyz-output to html converter because Pandoc converts for me.

3) I could let team members work in whatever format they want (e.g. Word) and I could include their content in a document via some kind of 
<pre>
```{=docx}
{{< include theword.docx >}}
```
</pre>
We hack processes like this now and this would be better.

**Reply by cscheid** [reply]
Thanks for the added context.

Compared to the table use case, I expect these to be quite a bit less common, and they offer a number of additional complications that I think we'd prefer not to support. (For example, if we're including theword.docx directly, we need to now worry about figure insertion, cross-reference resolution, etc.)

With that said, if you're willing to go the lua filter route, all of these are possible to implement by yourself: see https://pandoc.org/lua-filters.html#pandoc.read.

**Comment by rgaiacs** [comment]
I believe that I understand what @tbeason and @eeholmes want and why this is "hard" to implement.

They want to have Quarto/knitr/Jupyter generating the `.md` with

```
Some text

\begin{tabular}{ c c }
 cell1 & cell2 \\ 
 cell3 & cell4
\end{tabular}

More text
```

and Pandoc create the internal structure

```
[ Para [ Str "Some" , Space , Str "text" ]
, Table
    ( "" , [] , [] )
    (Caption Nothing [])
    [ ( AlignCenter , ColWidthDefault )
    , ( AlignCenter , ColWidthDefault )
    ]
    (TableHead ( "" , [] , [] ) [])
    [ TableBody
        ( "" , [] , [] )
        (RowHeadColumns 0)
        []
        [ Row
            ( "" , [] , [] )
            [ Cell
                ( "" , [] , [] )
                AlignDefault
                (RowSpan 1)
                (ColSpan 1)
                [ Plain [ Str "cell1" ] ]
            , Cell
                ( "" , [] , [] )
                AlignDefault
                (RowSpan 1)
                (ColSpan 1)
                [ Plain [ Str "cell2" ] ]
            ]
        , Row
            ( "" , [] , [] )
            [ Cell
                ( "" , [] , [] )
                AlignDefault
                (RowSpan 1)
                (ColSpan 1)
                [ Plain [ Str "cell3" ] ]
            , Cell
                ( "" , [] , [] )
                AlignDefault
                (RowSpan 1)
                (ColSpan 1)
                [ Plain [ Str "cell4" ] ]
            ]
        ]
    ]
    (TableFoot ( "" , [] , [] ) [])
, Para [ Str "More" , Space , Str "text" ]
]
```

instead of

```
[ Para [ Str "Some" , Space , Str "text" ]
, RawBlock
    (Format "tex")
    "\\begin{tabular}{ c c }\n cell1 & cell2 \\\\ \n cell3 & cell4\n\\end{tabular}"
, Para [ Str "More" , Space , Str "text" ]
]
```

In other words, they want Pandoc to also convert the LaTeX or TeX raw blocks created when reading the input as Markdown to `native`. It looks easy when you get a minimal working example but it can quickly get complex, for example the @tbeason's case of end a table line in latex with `\\[2mm]` instead of `\\`.

The best solution is the lua filter route suggested by @cscheid.

# Why does Pandoc process LaTeX equation but not LateX tables in Markdown?

Render math in the web has being challenge since the creation of the web. W3C addopted [MathML](https://www.w3.org/Math/) but all browsers didn't implemented it. The solution that people used for years was to have authors writing the equations in LaTeX and the server processing the LaTeX equations and create images. Recently, [MathJax](https://www.mathjax.org/) allowed LaTeX equations to be processed in the reader's web browser.

Because of this, Pandoc's implements "special" support for LaTeX math part. For example, consider the following `.md` file

```
Some text with inline math $x$ before a displayed math.

\begin{equation}
1 + 1
\end{equation}

More text without math.
```

The output of `pandoc -f markdown -t latex math.md` is

```
Some text with inline math \(x\) before a displayed math.

\begin{equation}
1 + 1
\end{equation}

More text without math.
```

Pandoc will keep all LaTeX math elements.

Now compare it to the output of `pandoc -f markdown -t html math.md`

```
<p>Some text with inline math <span
class="math inline"><em>x</em></span> before a displayed math.</p>
<p></p>
<p>More text without math.</p>
```

Note that Pandoc removed the display (`\begin{equation}`) math element. This is because of the `native` structure:

```
[ Para
    [ Str "Some"
    , Space
    , Str "text"
    , Space
    , Str "with"
    , Space
    , Str "inline"
    , Space
    , Str "math"
    , Space
    , Math InlineMath "x"
    , Space
    , Str "before"
    , Space
    , Str "a"
    , Space
    , Str "displayed"
    , Space
    , Str "math."
    ]
, Para
    [ RawInline
        (Format "tex") "\\begin{equation}\n1 + 1\n\\end{equation}"
    ]
, Para
    [ Str "More"
    , Space
    , Str "text"
    , Space
    , Str "without"
    , Space
    , Str "math."
    ]
]
```

Pandoc will use raw TeX block for the display math element. Today, most projects use MathJax, including Quarto, and the call to Pandoc is `pandoc -f markdown -t html --mathjax math.md` which produces

```
<p>Some text with inline math <span class="math inline">\(x\)</span>
before a displayed math.</p>
<p><span class="math display">\[\begin{equation}
1 + 1
\end{equation}\]</span></p>
<p>More text without math.</p>
```

With `--mathjax`, Pandoc goes into the raw TeX block and "extracts" the math portion of it and format it for MathJax (see the `\[` and `\]`).

**Comment by tarleb** [comment]
We had a similar issue [in a JOSS submission](https://github.com/openjournals/joss-reviews/issues/4442). Our solution was to parse *all* raw LaTeX blocks and inlines into pandoc's internal document format, except when we converting to/via LaTeX. I've published the filter as a quarto extension: [tarleb/parse-latex](https://github.com/tarleb/parse-latex). If I understand the discussion correctly, then this could be sufficient to resolve the issue:

``` bash
quarto install extension tarleb/parse-latex
```

Then in the `.qmd` document add this to the YAML frontmatter:

``` yaml
filters:
  - parse-latex
```

**Reply by cscheid** [reply]
This looks excellent!

**Reply by eeholmes** [reply]
Thanks! This will be super helpful in other contexts too. 

Alas it doesn't solve you original question of how to preserve the cross-references, no?

**Reply by eeholmes** [reply]
Hmm, I'd be nice if something like the below worked for html and latex. Maybe the filter can be modified to get this to work.

To my mind the
<pre>```{=latex}
</pre> 
syntax makes more sense as it is not just a div with class, but it is a code block. I am saying, "this is a type of code block and I want to use the commands (chunk yaml `#| `) that are available with code blocks." `:::{=xyz}` says to me "this is a div. apply whatever styling is appropriate for that".  But that's just my personal feeling.


```
```{=html}
#| label: tbl-one
#| tbl-cap: This is table one.
<table>
  <tr>
    <th>Company</th>
    <th>Contact</th>
    <th>Country</th>
  </tr>
  <tr>
    <td>Alfreds Futterkiste</td>
    <td>Maria Anders</td>
    <td>Germany</td>
  </tr>
  <tr>
    <td>Centro comercial Moctezuma</td>
    <td>Francisco Chang</td>
    <td>Mexico</td>
  </tr>
</table>
```

**Reply by cscheid** [reply]
@eeholmes Under that syntax, that's not behavior we can control, because it's actually Pandoc syntax. We choose not to override Pandoc syntax since that would be extremely confusing for Pandoc users.

**Reply by cscheid** [reply]
@eeholmes I will note that in the specific case of _html tables_, quarto _does_ convert them to latex:

````
---
format: pdf
keep-tex: true
---

A latex table in a LaTeX document:

```{=html}
<table>
  <tr>
    <th>Company</th>
    <th>Contact</th>
    <th>Country</th>
  </tr>
  <tr>
    <td>Alfreds Futterkiste</td>
    <td>Maria Anders</td>
    <td>Germany</td>
  </tr>
  <tr>
    <td>Centro comercial Moctezuma</td>
    <td>Francisco Chang</td>
    <td>Mexico</td>
  </tr>
</table>
```
````

<img width="855" alt="image" src="https://user-images.githubusercontent.com/285675/186717113-7234ac33-cda2-4543-9b78-5bbf294eaa24.png">

Note that this is a case in my "this is extremely confusing for Pandoc users" point!

We're probably going to improve the behavior here so that it's generally applicable to any format, and uses a less-confusing syntax.

**Reply by eeholmes** [reply]
Completely unrelated comment, but it is weird to me that 
<pre>
```{python}
</pre>
works but 
<pre>
```{html}
```{latex}
</pre>
do not. I keep wanting to do something like
<pre>
```{html}
#| div: class="abc"
<a bunch of html>
</pre>
because I've gotten used to `#| ` as a way to apply attributes to a code chunk and I think of html and latex as different types of code.

**Reply by cscheid** [reply]
The mnemonic to keep in mind is that braces without the equals sign means _computational content_: stuff that needs to be executed by another program and that generates an output: so those go for things like

````

```{r}
```

```{julia}
```

```{python}
```

```{ojs}
```

````

In addition, these are the only blocks that understand hash-pipe syntax.

braces _with_ the equals sign mean "pass this as-is to the specific Pandoc format" (aside from that special case in html tables, which I still contend is very confusing behavior on the part of quarto!)

**Comment by eeholmes** [comment]
@cscheid It was the cross-reference part that is not being used. The OP's question was about cross-references. I know there are probably other ways to do this for html but not for latex.

<pre>
```{=html}
#| label: tbl-one
#| tbl-cap: This is table one.
</pre>

My point here was that, "oh, if this cross-ref syntax worked for html, then it should be easy to get this to work for latex by extending somehow the OP's lua filter."

if I do this

<pre>
```{r}
#| label: tbl-one
#| tbl-cap: This is table one.
--code that outputs html or latex--
```
</pre>

Then the cross-reference works, but it doesn't work if I put the html or latex there raw.

**Reply by eeholmes** [reply]
Which suggests a very hacky solution of creating an R function (or Python or Julia) that just inputs raw latex and then output a table object that `{r}` or `{python}` already recognizes.

I think I have done this with {xtable} in R when I was fighting with table output to HTML and LaTex once.

**Reply by cscheid** [reply]
(we're getting sprawled over subthreads here, so I apologize for the fractured replies!)

As I mentioned in my [other comment](https://github.com/quarto-dev/quarto-cli/discussions/2052#discussioncomment-3476351), the mnemonic to keep in mind is that 'equals inside braces as a code block specifier means "Pandoc will take this as-is"', so there no place for the `#|` to get interpreted (those only work in computational content, which `html` isn't).

If this feels a little confusing, that's (in the personal opinion of this quarto developer!) because it kind of is. We are going to substantially improve the situation for cross-references, captions, and labels for v1.2, though, so stay tuned!

**Reply by cscheid** [reply]
@eeholmes You got it: in order for your trick to work, you would have to hack it through the computational content code cells.

**Reply by eeholmes** [reply]
Yes, I think of `{latex}` as opposed to `{=latex}` exactly as you wrote. So when I use `{latex}`, I am saying that I don't want Pandoc to take "as-is". I want this put through an interpreter first (computational or otherwise, but just not as-is). Same with `{html}` as opposed to `{=html}`. The first is "hey plug this into the interpreter so I can have `#| ` and other good stuff applied" and the second is "please treat this as-is and pass it through".  Actually same with `{python}` and `{=python}` for example. One is sent to an engine, the other is formatted as raw code but not evaluated. *Note, I am just thinking out-loud as to what feels intuitive to me personally.*

**Reply by cscheid** [reply]
I totally understand! I'm just trying to give as much context to the current state of quarto and our decisions as I can.

With that said, here's how you'd make it all work, although it's completely atrocious:

````
---
format: pdf
keep-tex: true
---

A latex table in a LaTeX document:

## An attempt that doesn't work

```{=html}
<table>
  <tr>
    <th>Company</th>
    <th>Contact</th>
    <th>Country</th>
  </tr>
  <tr>
    <td>Alfreds Futterkiste</td>
    <td>Maria Anders</td>
    <td>Germany</td>
  </tr>
  <tr>
    <td>Centro comercial Moctezuma</td>
    <td>Francisco Chang</td>
    <td>Mexico</td>
  </tr>
</table>
```

: A table with a caption

## An attempt that does

| fruit  | price  |
|--------|--------|
| apple  | 2.05   |
| pear   | 1.37   |
| orange | 3.09   |

: A table with a caption

## An atrocious hack that combines both

```{r}
#| tbl-cap: A table with a caption
#| output: asis
#| echo: false
cat(paste(
"```{=html}",
"<table>",
"  <tr>",
"    <th>Company</th>",
"    <th>Contact</th>",
"    <th>Country</th>",
"  </tr>",
"  <tr>",
"    <td>Alfreds Futterkiste</td>",
"    <td>Maria Anders</td>",
"    <td>Germany</td>",
"  </tr>",
"  <tr>",
"    <td>Centro comercial Moctezuma</td>",
"    <td>Francisco Chang</td>",
"    <td>Mexico</td>",
"  </tr>",
"</table>",
"```", sep="\n"))
```
````

That .qmd produces this output:

<img width="532" alt="image" src="https://user-images.githubusercontent.com/285675/186722671-ef04c2fc-fb89-4114-91c8-b0262f88aa47.png">

**Reply by cscheid** [reply]
@tbeason If you truly, absolutely need it right now, look at the above code :) But if you can wait two or three months, we are working on a much better solution for v1.2.

**Reply by eeholmes** [reply]
Ugly hack but shows how it can be done. For R users, {xtable} has a LaTeX to html function. So you can read the latex into a string, use the xtable latex to html converter function and then the hack above.

**Reply by tbeason** [reply]
Do I understand correctly, though, that the `parse-latex` filter/extension mentioned above would make this cross-reference work if I just write the latex table in pure latex (including the caption and label)? Because then pandoc would see the label? Or would it still fail? Sorry no time to test it right now.

**Reply by tarleb** [reply]
@tbeason that exact case (i.e., cross-referencing of tables) was the original motivation for the filter. So, yes, it *should* work for that. (But it's still software, so I'll stay with the conjunctive mood.)

**Comment by eeholmes** [comment]
Here is a test of @tarleb 's filter. @tbeason 

<pre>
---
title: "test"
format: 
  html: default
  pdf:
    keep-tex: true
editor: source
filters:
  - parse-latex
---

This is a reference to Table @tbl-one.


```{=latex}
\begin{table}
\caption{caption of table.}\label{tbl-one}%
\begin{center}
\begin{tabular}{ c c c }
 cell1 & cell2 & cell3 \\ 
 cell4 & cell5 & cell6 \\  
 cell7 & cell8 & cell9    
\end{tabular}
\end{center}
\end{table}
```
</pre>


PDF output: Note if I use `\ref{tbl-one}` instead the ref show up in PDF but not in HTML.

<img width="555" alt="image" src="https://user-images.githubusercontent.com/2545978/186750870-43902f5f-eba2-4e3f-a166-b7180843b93f.png">



HTML output:

<img width="1133" alt="image" src="https://user-images.githubusercontent.com/2545978/186750586-b5c7101a-b66e-48f5-b0cf-7e33e2ea4a20.png">

**Reply by tarleb** [reply]
:disappointed: But now that I think about it some more, the result makes sense, given that the LaTeX is not parsed into the internal format when converting to PDF. I forgot that the cross-referencing system that we use for JOSS is a bit different. I'll check tomorrow, maybe I can make it work in Quarto.

**Reply by eeholmes** [reply]
Do you know why it added the caption as "(a) caption of table" instead of adding to "Table 1: caption of table"

**Reply by tarleb** [reply]
I don't really understand what's going on there. It seems to be some weird interaction of Quarto and pandoc's LaTeX parser, more specifically their use of Div wrappers.

**Reply by tarleb** [reply]
It seems that Quarto expects the ID to be encoded in the table's caption and gets confused if the table has an ID in the pandoc AST.
We get better results if we change the `RawBlock` function to this hack:

``` lua
function RawBlock (raw)
  if raw.format:match 'tex' then
    return pandoc.read(raw.text, 'latex').blocks:walk {
      Div = function (div)
        local tbl = div.content[1]
        if #div.content == 1 and tbl.t == 'Table' then
          local idstr = string.format('{#%s}', div.identifier)
          if #tbl.caption.long == 0 then
            tbl.caption.long = pandoc.Blocks{pandoc.Plain(idstr)}
          else
            tbl.caption.long[1].content:insert(idstr)
          end
          return tbl
        end
      end,
    }
  end
end
```

**Reply by tarleb** [reply]
The best workaround that I found was to use LaTeX syntax like `\autoref{tbl-one}` in combination with this abomination:

``` lua
if FORMAT:match 'latex' then
  return {}
end

function RawBlock (raw)
  if raw.format:match 'tex' then
    return pandoc.read(raw.text, 'latex').blocks:walk {
      Div = function (div)
        local tbl = div.content[1]
        if #div.content == 1 and tbl.t == 'Table' then
          local idstr = string.format('{#%s}', div.identifier)
          if #tbl.caption.long == 0 then
            tbl.caption.long = pandoc.Blocks{pandoc.Plain(idstr)}
          else
            tbl.caption.long[1].content:insert(idstr)
          end
          return tbl
        end
      end,
    }
  end
end

function RawInline(raw)
  if raw.format:match 'tex' then
    return pandoc.utils.blocks_to_inlines(
      pandoc.read(raw.text, 'latex').blocks:walk {
        Link = function (link)
          local ref = link.attributes.reference
          if ref then
            return pandoc.Cite(
              '@' .. ref,
              {pandoc.Citation(ref, 'AuthorInText')}
            )
          end
        end
      }
    )
  end
end
```

I'm not comfortable with actually using something like that, but it does the job: Both the HTML and PDF output look ok.

**Reply by tbeason** [reply]
Hmm yea not ideal. I am not sure what I will do. I guess temporarily I can just remember which tables I need to reference using raw latex. 

Could there be a different solution? For example, if Quarto cannot find the reference for `@tbl-one` in this case, maybe instead of turning it into `\textbf{?tbl-one}` it should default to `\ref{tbl-one}`? For PDF/latex it would be fine I think because there are 2 possibilities: (1) the label actually exists or (2) it doesn't, in which case latex handles the missing reference itself. Right?
